server/contexts/ctxerr/ctxerr.go

```diff
@@ -13,87 +13,190 @@ package ctxerr
+func (e *FleetError) Unwrap() error {
+	return e.cause
+}
+
+// MarshalJSON implements the marshaller interface, giving us control on how
+// errors are json-encoded
+func (e *FleetError) MarshalJSON() ([]byte, error) {
+	return json.Marshal(fleetErrorJSON{
+		Message: e.msg,
+		Data:    e.data,
+		Stack:   e.stack.List(),
+	})
+}
+
+// New creates a new error with the given message.
+func New(ctx context.Context, msg string) *FleetError {
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, nil, nil}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// NewWithData creates a new error with the given message and attaches
+// aditional data to it.
+func NewWithData(ctx context.Context, msg string, data map[string]interface{}) *FleetError {
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, nil, data}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// Errorf creates a new error with the given message.
+func Errorf(ctx context.Context, format string, args ...interface{}) *FleetError {
+	msg := fmt.Sprintf(format, args...)
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, nil, nil}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// Wrap creates a new error with the given message, wrapping another error.
+func Wrap(ctx context.Context, cause error, msgs ...string) *FleetError {
+	msg := strings.Join(msgs, " ")
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, cause, map[string]interface{}{}}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// WrapWithData creates a new error with the given message, wrapping another
+// error and attaching the data provided to it.
+func WrapWithData(ctx context.Context, cause error, msg string, data map[string]interface{}) *FleetError {
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, cause, data}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// Wrapf creates a new error with the given message, wrapping another error.
+func Wrapf(ctx context.Context, cause error, format string, args ...interface{}) *FleetError {
+	msg := fmt.Sprintf(format, args...)
+	stack := NewStack(1)
+	err := &FleetError{msg, stack, cause, nil}
+	ensureCommonMetadata(ctx, err)
+	return err
+}
+
+// Cause returns the root error in err's chain.
+func Cause(err error) error {
 	for {
-		uerr := errors.Unwrap(err)
+		uerr := Unwrap(err)
 		if uerr == nil {
-			if errors.As(err, &cerr) {
-				uerr = cerr.Cause()
-			} else {
-				break
-			}
+			return err
 		}
 		err = uerr
 	}
-
-	return err
 }
 
-// NewContext returns a context derived from ctx that contains the provided
-// error handler.
-func NewContext(ctx context.Context, eh *errorstore.Handler) context.Context {
-	return context.WithValue(ctx, errHandlerKey, eh)
+// Unwrap is a wrapper of built-in errors.Unwrap. It returns the result of
+// calling the Unwrap method on err, if err's type contains an Unwrap method
+// returning error. Otherwise, Unwrap returns nil.
+func Unwrap(err error) error {
+	return errors.Unwrap(err)
```

Comment:
the idea was to provide our own wrapper like we do with the rest, in case we want to change the underlying implementation again in the future. Honestly not a big deal if we decide to remove it.

Reply:
Yeah it's true that we have `Errorf`, `New` and all that, it could be nice not to have to import the stdlib `errors` package elsewhere in many cases.